src/structure.js

```diff
@@ -169,13 +169,19 @@ Structure.prototype.reference = function (path) {
   if (isCursor(path) && path._keyPath) {
     path = path._keyPath;
   }
-  var self = this, pathId = pathString(path);
-  var listenerNs = self._pathListeners[pathId];
-  var cursor = this.cursor(path);
 
-  var changeListener = function (newRoot, oldRoot, changedPath) { cursor = self.cursor(path); };
-  var referenceListeners = [changeListener];
-  this._pathListeners[pathId] = !listenerNs ? referenceListeners : listenerNs.concat(changeListener);
+  path = valToKeyPath(path) || [];
+
+  var self = this,
+      listenerNs = getListenerNs(self._pathListeners, path),
+      cursor = this.cursor(path),
+      changeListener = function() {
+        cursor = self.cursor(path);
+      };
+
+  if (!listenerNs[0]) {
+    listenerNs.push(changeListener);
```

Comment:
Right now this whole thing was built without consideration of using anything other than key paths that look like ['foo', 'bar', 'baz'], an array of strings. getListenerNs calls concat on the provided path, adding in the '__listeners__'  path, this array is passed to deepGet, if it has already been added it would be an array because if it is not added we set it as an array. If you're worried about potential collisions with users using that string in their code you can use some generated random string of sufficient length instead.